### PR TITLE
issue #593 - adding simple auto-scroll checkbox to enable and disable

### PR DESCRIPTION
**Pull request by bfitzpat**
Signed-off-by: bfitzpat@redhat.com <bfitzpat@redhat.com>

### CODE REVIEW

**Comment by bfitzpat** [comment]
Though this may not be the best solution to this issue, it provides a simple solution that may be enough to address the scrolling problem. It's something we came up with locally and thought you might find useful.

File: src/components/logs/logsWebview.ts [hunk 1]
```diff
@@ -80,6 +82,12 @@ export class LogsPanel extends WebPanel {
                     });
                     // TODO: need to apply filters here!
                     elt.appendChild(document.createTextNode(message.text));
+
+                    // handle auto-scroll on/off
+                    var checkBox = document.getElementById("autoscroll");
+                    if (checkBox.checked == true){
+                        document.getElementById('bottom').scrollIntoView();
```
Comment (itowlson) [inline src/components/logs/logsWebview.ts]: This element doesn't seem to be present in the HTML.  I found if I added a `<p id='bottom'></p>` after the `<code>...</code>` around line 59-62 then it worked, but that was only really simple testing...

File: src/components/logs/logsWebview.ts [hunk 2]
```diff
@@ -51,6 +51,8 @@ export class LogsPanel extends WebPanel {
                 </select>
                 <span style='position: absolute; left: 240px'>Match expression</span>
                 <input style='left:350px; position: absolute' type='text' id='regexp' onkeyup='eval()' placeholder='Filter' size='25'/>
+                <span style='position: absolute; left: 600px'>Auto-scroll</span>
+                <input style='left: 675px; position: absolute' type='checkbox' id='autoscroll' onkeyup='eval()' checked="true"/>
```
Comment (itowlson) [inline src/components/logs/logsWebview.ts]: This makes the checkbox appear even in Show Logs (non-tailing) mode.  I'm not sure this makes sense.  Could we hide the caption and check box in that mode?  (The way it's done at the moment, the panel doesn't know whether the user is showing or following, so we'd need the controlling command to pass that info in I guess.)

File: src/components/logs/logsWebview.ts (hunk 1 above)
Comment (bfitzpat) [reply src/components/logs/logsWebview.ts]: Sorry about missing that in my first pass, but I have added it back in.

File: src/components/logs/logsWebview.ts (hunk 2 above)
Comment (bfitzpat) [reply src/components/logs/logsWebview.ts]: We just have it showing up in all cases currently. There wasn't a clear way to identify when we were coming in via show or follow.

**Comment by bfitzpat** [comment]
If the log is extremely long, it seems to get hung up sometimes (I think due to the code in eval() it has issues processing super-long logs at times), but otherwise it works great as a very simple solution.

**Comment by lstocchi** [comment]
Played with it and it worked fine for me but i have a doubt...by reading https://github.com/Azure/vscode-kubernetes-tools/issues/593#issuecomment-516672587 it looks like this doesn't cover what we really need. If you scroll up and a new log is written it takes you to the bottom. It can be quite frustrating.
Why don't we listen to `scroll`? Has it been tested before? Are there any issues?
My idea is this.

Have a variable that stores a boolean which specify if the user is at the end of the page. If true the window need to stick to the bottom when new logs are added, if false it means user scroll up.
```
let isToBottom = true;   
if (isToBottom) document.getElementById('bottom').scrollIntoView();
```

To set the variable we listen to the scroll event
```
<div id="logPanel" style="overflow-y: scroll; height: 100%">
```
```
const elem = document.getElementById('logPanel');
let lastScrollTop = 0;
const toBottom = debounce(function() {
        const st = elem.scrollTop;
        if (st > lastScrollTop){
             // scroll down
             isToBottom = (elem.scrollTop + window.innerHeight) >= elem.scrollHeight;
        } else {
             // scroll up
            isToBottom = false;
         }
            lastScrollTop = st <= 0 ? 0 : st; 
  }, 250);

elem.addEventListener("scroll", toBottom);
```
where the debounce function is taken from https://davidwalsh.name/javascript-debounce-function 
so we don't always fire the function.

```
// Returns a function, that, as long as it continues to be invoked, will not
// be triggered. The function will be called after it stops being called for
// N milliseconds. If `immediate` is passed, trigger the function on the
// leading edge, instead of the trailing.
function debounce(func, wait, immediate) {
	var timeout;
	return function() {
		var context = this, args = arguments;
		var later = function() {
			timeout = null;
			if (!immediate) func.apply(context, args);
		};
		var callNow = immediate && !timeout;
		clearTimeout(timeout);
		timeout = setTimeout(later, wait);
		if (callNow) func.apply(context, args);
	};
};
```

If this all works we shouldn't need the checkbox but we should have a `go to the bottom` button so if there are many logs one doesn't have to scroll down for hours...

**Comment by itowlson** [comment]
@lstocchi This sounds like a good approach - are you able to take it on, or @bfitzpat do you want to take a crack at Luca's suggestion?

**Comment by bfitzpat** [comment]
User cases discussed in meeting:
1. User-specified path to binary. (don't have currently)
2. System Path-specified binary.
3. If neither 1 or 2, download latest released version from github.

**Comment by bfitzpat** [comment]
> @lstocchi This sounds like a good approach - are you able to take it on, or @bfitzpat do you want to take a crack at Luca's suggestion?

I can take a look.

**Comment by lstocchi** [comment]
Hey @bfitzpat, i noticed a couple of errors.
First is `window.addEventListener("scroll", toBottom);`. It should be `elem.addEventListener("scroll", toBottom)`. The scroll is inside that div, not in the window.
The second (my bad) when creating the `toBottom` function it is saved inside a constant `const toBottom = debounce(function() {` but the debouce function returns a function so it must be `let toBottom = debounce(function() {`
After these changes it should be fine.

**Comment by bfitzpat** [comment]
@lstocchi that's awesome. Those two fixes and we were off to the races. Thanks! 

I'll see about adding a button that scrolls to the bottom from the tools area at the top of the log window next!

**Comment by lstocchi** [comment]
I guess this is done. I tested it and it works, i couldn't test it with all different use cases i had in mind because there are other PRs still opened related to the logView but this looks good to me. @itowlson WDYT?

I don't know why travis still shows it's pending but it's finished

**Comment by lstocchi** [comment]
@itowlson Tested it with the new release. Works fine!! Thank you.

Thanks @bfitzpat for your contribution :+1:

**Pull request merged**